### PR TITLE
Horizontal Nav with dot notification

### PR DESCRIPTION
**Pull request by imchristian**
Now the horizontal navigation options allow specifying whether we need to show a notification dot or not for a given section.

DotNotificationScreenshot

+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] `UNRELEASED.md`
+ [ ] `README.md`
+ [ ] Tests
+ [ ] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/master/Backpack/Backpack.h)
+ [ ] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m [hunk 1]
```diff
@@ -32,18 +32,27 @@ @interface BPKHorizontalNavigationItemDefault ()
 @property(readonly) UIColor *contentColor;
 @property(nonatomic) CGFloat horizontalSpacing;
 @property(nonatomic) CGFloat verticalSpacing;
+@property(nonatomic) BOOL showNotificationDot;
+@property(nonatomic, strong) UIView *notificationDot;
```
Comment (k0nserv) [inline Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: This should be `readonly` right?

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m [hunk 2]
```diff
@@ -186,5 +197,30 @@ - (CGFloat)horizontalSpacing {
...
+
+        [self addSubview:_notificationDot];
+        [NSLayoutConstraint activateConstraints:@[
+            [_notificationDot.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-self.horizontalSpacing],
+            [_notificationDot.topAnchor constraintEqualToAnchor:self.topAnchor],
+            [_notificationDot.heightAnchor constraintEqualToConstant:BPKSpacingMd],
+            [_notificationDot.widthAnchor constraintEqualToConstant:BPKSpacingMd]
+        ]];
+
+        _notificationDot.layer.cornerRadius = BPKSpacingMd/2;
```
Comment (k0nserv) [inline Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: Nitpic: Spacing around `/`

File: Example/Backpack/View Controllers/HorizontalNavViewController.swift [hunk 3]
```diff
@@ -27,15 +27,25 @@ class HorizontalNavViewController: UIViewController, BPKTappableLinkLabelDelegat
...
 
         if showIcons {
             horizontalNav.options = [
-                BPKHorizontalNavigationOption(name: "Flights", iconName: .flight, tag: 0),
-                BPKHorizontalNavigationOption(name: "Hotels", iconName: .hotels, tag: 1),
-                BPKHorizontalNavigationOption(name: "Car hire", iconName: .cars, tag: 2)
+                BPKHorizontalNavigationOption(name: "Flights",
+                                              iconName: .flight,
+                                              tag: 0,
+                                              showNotificationDot: showNotificationDot),
```
Comment (k0nserv) [inline Example/Backpack/View Controllers/HorizontalNavViewController.swift]: Shouldn't only one of the tabs show the dot? Ideally the first one that's not selected i.e. `Hotels`?

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m [hunk 4]
```diff
@@ -176,6 +186,7 @@ - (void)setup {
     self.titleLabel.lineBreakMode = NSLineBreakByTruncatingTail;
     [self updateStyle];
     [self updateIconStyle];
+    [self showNotificationDot:self.showNotificationDot];
```
Comment (georgegillams) [inline Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: this name seems confusing as it could be used to hide the dot. Maybe 
```suggestion
    [self updateNotificationDot:self.showNotificationDot];
```

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationOption.m [hunk 5]
```diff
@@ -77,6 +78,23 @@ - (instancetype)initWithName:(NSString *)name iconName:(BPKIconName)iconName tag
     return self;
 }
 
+- (instancetype)initWithName:(NSString *)name
```
Comment (georgegillams) [inline Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationOption.m]: Could `initWithName:(NSString *)name iconName:(BPKIconName)iconName tag:(NSInteger)tag` just call this method now?

**Comment by georgegillams** [comment]
Is it possible to update the dot at runtime without changing the options?
If it is, we should probably have a story for that in the example app

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m (hunk 1 above)
Comment (imchristian) [reply Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: It's in the .m file so 🤷‍♂️

Comment (k0nserv) [reply Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: It's still helpful though? It expresses the intent better for anyone reading the code even if it has no impact on consumers?

Comment (imchristian) [reply Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: Yeah that's fair I'll do the change 👍

**Comment by imchristian** [comment]
@georgegillams as we've discussed yesterday through slack there's not actually a way to update an option without replacing it entirely and since the feature to update the dot icon without replacing the whole option is not a requirement now we decided to not implement it.

**Comment by imchristian** [comment]
@k0nserv I've changed the constraint logic, I didn't try to constraint it to the `UILabel` before because I thought that I'll be unreliable but it's actually working pretty well so good catch 👍 
The snapshot tests are still showing changes even though I deleted the `sizeToFit` but as far as I can see there isn't any visual change between both screenshots.

**Comment by k0nserv** [comment]
There is, look at `testNarrow_light-mode@2x.png` for example

**Comment by imchristian** [comment]
@k0nserv you're right, I'll investigate more because I don't know what could cause that

**Comment by imchristian** [comment]
Ok, so here are the learnings:
The first approach was, of course, follow with the `UIEdgeInsetsMake` + `CGRect` usage to positioning the dot subview inside the `UIButton`.
This approach was extremely unreliable and it wasn't always in the best position (see @k0nserv [suggestion](https://github.com/Skyscanner/backpack-ios/pull/609#pullrequestreview-374299107))

After that, I tried with auto layout to positioning the dot subview, it's working perfectly to options without icons since the auto layout is renderer properly. The problems came up when you add a dot notification using also the inner `UILabel` and `UIImageView` in the `UIButton` 🤔
It was adding an extra space (16pt) between the icon and label, why? well I went to the books to find out this black magic problem:

**"Autolayout is implemented through the superview chain, so if a view uses autolayout, then automatically so do all its superviews"**

Basically when we add with autolayout the dot subview all the `UIButton` start to render their own constraints.
Those are not cancellable using `autoresizingMaskIntoConstraint` 🤷‍♂️ and because of that we cannot render the whole `UIButton` with our own constraints.

It's not a problem for the use case that we need now which is the section with title and dot notification with no icons, so the icon-label padding is not happening.
To avoid that we've limited through the init the dot notification to be used only with title and without icons.

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.h [hunk 6]
```diff
@@ -55,7 +55,17 @@ NS_SWIFT_NAME(HorizontalNavigationItemDefault) IB_DESIGNABLE @interface BPKHoriz
...
+- (instancetype)initWithName:(NSString *)name iconName:(BPKIconName)iconName;
+
+/**
+ * Create a `BPKHorizontalNavigationItem` with a set of options and an optionaly selected option.
+ *
+ * @param name NSString the name for the navigation item
+ * @param iconName BPKIconName the icon for the navigation item
+ * @param showDot send YES to show a red notification dot
+ */
+- (instancetype)initWithName:(NSString *)name iconName:(BPKIconName)iconName showNotificationDot:(BOOL)showDot NS_DESIGNATED_INITIALIZER;
```
Comment (georgegillams) [inline Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.h]: I think the public interface should not have a method for initialising with an icon and dot if this is unsupported. It might still be good to have this method privately for the others to call?

Comment (k0nserv) [reply Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.h]: I think that's supported? What's not supported it usin git with a badge?

**Comment by k0nserv** [comment]
> It's not a problem for the use case that we need now which is the section with title and dot notification with no icons, so the icon-label padding is not happening.
To avoid that we've limited through the init the dot notification to be used only with title and without icons.

Can we not  use manual layout to position the dot to get around this?

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.h (hunk 6 above)
Comment (imchristian) [reply Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.h]: Well, this class is [not declared](https://github.com/Skyscanner/backpack-ios/blob/master/Backpack/HorizontalNavigation/Classes/HorizontalNavigation.h) to be used directly so that's why I didn't think twice this initializer but we can still discuss if it's necessary anyway to do that change

**Comment by imchristian** [comment]
@k0nserv by the manual layout you mean not to specify the positions using `UIEdgeInsetsMake` for the option with icon and title?
If that's what you mean I think it's not the right path to follow since Apple recommends to use `self.imageEdgeInsets` and `self.titleEdgeInsets` to change a native `UIButton` layout and as I said on my comment if we start adding constraints we can't deactivate the ones added by Apple with `translatesAutoresizingMaskIntoConstraints`.

**Comment by k0nserv** [comment]
Well use `imageEdgeInsets` and `titleEdgeInsets` to position the title and icon, then use `layoutSubviews` to layout the dot only

**Comment by imchristian** [comment]
@k0nserv I think that I'm not following you perhaps we should move to a pair session to tackle this down FYI @georgegillams

**Comment by k0nserv** [comment]
Sounds good to me

**Comment by imchristian** [comment]
Screenshots updated with the latest changes FYI @georgegillams @k0nserv 

Simulator Screen Shot - iPhone 8 - 2020-03-16 at 12 05 38
Simulator Screen Shot - iPhone 8 - 2020-03-16 at 12 06 02

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m [hunk 7]
```diff
@@ -32,19 +32,32 @@ @interface BPKHorizontalNavigationItemDefault ()
 @property(readonly) UIColor *contentColor;
 @property(nonatomic) CGFloat horizontalSpacing;
 @property(nonatomic) CGFloat verticalSpacing;
+@property(nonatomic, readonly) UIView *notificationDot;
 
 @end
 
 @implementation BPKHorizontalNavigationItemDefault
 
+@synthesize notificationDot = _notificationDot;
```
[the remark this answers is not in the document]
Comment (k0nserv) [reply Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationItemDefault.m]: I checked locally by removing it and everything is fine

File: Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationOption.h [hunk 8]
```diff
@@ -88,5 +88,30 @@ NS_ASSUME_NONNULL_BEGIN
 */
 - (instancetype)initWithName:(NSString *)name iconName:(BPKIconName)iconName tag:(NSInteger)tag;
 
+/**
+* Intialize an option with a given name, tag, and icon.
+*
+* @param name The name of the option.
+* @param tag The tag of the option.
+* @param showDot send YES to show a red notification dot
+* @return A named option with an icon and a tag.
```
Comment (k0nserv) [inline Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigationOption.h]: This is outdated
